File: FSharpConsoleApp/HappyPathCoding.fs
module FSharpConsoleApp.HappyPathCoding

open FsToolkit.ErrorHandling

type Customer =
    {
        Id : int
        Name : string option
    }

type Data =
    {
        Id : int
        Amount : int
    }

let loadCustomer customerId =
    asyncResult {
        do! customerId = 42
            |> Result.requireTrue
                   "customer not found"
        return
            {
                Customer.Id = customerId
                Name = Some "Charles"
            }
    }

let loadCustomer' customerId =
    asyncResult {
        return! Error "customer not found"
    }

let loadCustomerCaller () =
    async {
        let! customerResult = loadCustomer 42

        return
            match customerResult with
            | Ok customer -> "we have a customer"
            | Error error -> "we have no customer"
    }

let loadData dataId =
    asyncResult {
        return
            {
                Data.Id = dataId
                Amount = 100
            }
    }

let getNameOfCustomer customer =
    option {
        let! name = customer.Name
        return name
    }

let getCustomerNameAndAmount customerId dataId =
    asyncResult {
        let! customer = loadCustomer customerId
        let! data = loadData dataId

        let! name =
            customer
            |> getNameOfCustomer
            |> Result.requireSome
                   "customer has no name"

        return name, data.Amount
    }

let compute () =
    async {
        let! result = getCustomerNameAndAmount 42 17
        match result with
        | Ok (name, amount) -> printf $"customer = {name}, amount = {amount}"
        | Error error -> printf $"error is {error}"
    }

File: CSharpConsoleApp/HappyPathCoding.cs
using System;
using System.Threading.Tasks;

namespace CSharpConsoleApp
{
    public class HappyPathCoding
    {
        public record Customer(int Id, string? Name);
        public record Data(int Id, int Amount);

        private async Task<Customer?> GetCustomer(
            int customerId)
        {
            return new(customerId, "Charles");
        }

        private async Task<Data?> GetData(
            int dataId)
        {
            return new(dataId, 100);
        }

        private string? GetNameOfCustomer(
            Customer customer)
        {
            return customer.Name;
        }

        public async Task<(string Name, int Amount)?>
            GetCustomerNameAndAmount(
                int customerId, int dataId)
        {
            var customer = await GetCustomer(customerId)
                .ConfigureAwait(false);

            if (customer == null)
            {
                return null;
            }

            var data = await GetData(dataId)
                .ConfigureAwait(false);

            if (data == null)
            {
                return null;
            }

            var name = GetNameOfCustomer(customer);

            if (name == null)
            {
                return null;
            }

            return (name, data.Amount);
        }

        public async Task Compute()
        {
            var result = await GetCustomerNameAndAmount(42, 17);
            if (result != null)
            {
                Console.WriteLine($"customer = {result.Value.Name}, amount = {result.Value.Amount}");
            }
            else
            {
                Console.WriteLine("error");
            }
        }
    }
}